### PR TITLE
Internal lint: suggest `is_type_diagnostic_item` over `match_type` where applicable

### PR DESCRIPTION
**Pull request by ebroto**
changelog: none

### CODE REVIEW

**Comment by mikerite** [comment]
This is a great idea.

Is there any reason to save the path to a diagnostic item at all? I think it might be easier to loop through the lines of `utils/path.rs` and apply the check there.

Maybe you could make a similar lint to check for lang items as well.

**Comment by ebroto** [comment]
>Is there any reason to save the path to a diagnostic item at all?

>Maybe you could make a similar lint to check for lang items as well.

Yeah my initial idea was to get rid of diagnostic items in the `paths` module, but it's not as straightforward as it could seem. For example, in some cases paths to diagnostic items are put together in a slice with paths to other types, and fed to `match_def_path`.

But you make me think that maybe we could work on the clippy `utils` module to expose functions that take a type that could be either a diagnostic item, a lang item or another type, using an enum, and avoid using the relevant rustc API functions directly.
I would prefer to discuss this design in a Zulip thread though (it seems we're moving there) and get some feedback before trying to implement it. 

If you agree I would go with merging this before, knowing it only addresses a part of the problem. I think it can be helpful to avoid missing it in the reviews.

File: clippy_lints/src/utils/internal_lints.rs [hunk 1]
```diff
@@ -652,3 +675,89 @@ fn suggest_note(
...
+            if let Some(item_name) = diag_items.iter().find_map(|(k, v)| if *v == ty_did { Some(k) } else { None });
+            then {
+                let cx_snippet = snippet(cx, context.span, "_");
+                let ty_snippet = snippet(cx, ty.span, "_");
+
+                span_lint_and_sugg(
+                    cx,
+                    MATCH_TYPE_ON_DIAGNOSTIC_ITEM,
+                    expr.span,
+                    "usage of `utils::match_type() on a type diagnostic item`",
```
Comment (giraffate) [inline clippy_lints/src/utils/internal_lints.rs:707]: nits
```suggestion
                    "usage of `utils::match_type()` on a type diagnostic item",
```

**Comment by mikerite** [comment]
> If you agree I would go with merging this before, knowing it only addresses a part of the problem. I think it can be helpful to avoid missing it in the reviews.

I think it's an good improvement as is. I'm happy with whatever you and the reviewer decide.

**Review by flip1995 — APPROVED** [review]
I agree that we can merge this as is. Doing the same with `LangItem`s would be great. Please address the comment by @giraffate and r=me after that.

**Pull request merged**